### PR TITLE
Test fuzzers

### PR DESCRIPTION
**Pull request by tokatoka**
I've made scripts for 6 fuzzers in fuzzer/ (excluding baby, generic, frida, fuzzbench), that can test the building for each,
so that we can tell if something is going wrong at CI.
yes they are bash scripts and the platform-independent workaround is a TODO.

what do you think 🤗 ?

### CODE REVIEW

**Comment by tokatoka** [comment]
Hm `CMake Error: Unknown argument --disable-shared` ?

**Comment by andreafioraldi** [comment]
disable shared is a configure argument, use configure

**Comment by tokatoka** [comment]
the scripts have to install a bunch of dependencies to build libmozjpeg & llibpng

**Comment by domenukk** [comment]
Maybe we should use the docker container for ci then?

**Comment by tokatoka** [comment]
Running the build_all_fuzzer.sh inside the container?
https://github.com/AFLplusplus/LibAFL/blob/5b54f0f0680763f2cbaf9831382bc0d7acf0e2d4/Dockerfile#L82

We can enable this line, but no idea how it can be integrated into ci. (I know little about Docker 😅 )

**Comment by tokatoka** [comment]
The current Dockerfile isn't used in build_and_test.yml?

**Comment by domenukk** [comment]
No it's not yet used 🙃

**Comment by tokatoka** [comment]
I'd like to wait for that then...

**Comment by domenukk** [comment]
We don't need the dockerfile for the rest though. Maybe just install the deps in a makefile or script for now?

**Comment by tokatoka** [comment]
Wierd, it says
` libc6-dev is already the newest version (2.31-0ubuntu9.2).`
but `configure` can't find these files
```
-- Looking for sys/types.h
-- Looking for sys/types.h - not found
-- Looking for stdint.h
-- Looking for stdint.h - not found
-- Looking for stddef.h
-- Looking for stddef.h - not found
-- Check size of size_t
-- Check size of size_t - failed
-- Check size of unsigned long
-- Check size of unsigned long - failed
```
🤷‍♂️

**Comment by tokatoka** [comment]
Oh, I got it. It's libafl_cc's problem like I mentioned in the discord chat yesterday.
We don't rely on any more dependencies on `ubuntu-latest`, I confirmed it in my personal repo
https://github.com/tokatoka/ci-debug/commit/a544e27931b31641328b00e6ef96f7d52a5c7d1f
Then I'll wait for it to be fixed in main branch

**Comment by domenukk** [comment]
This https://github.com/include-what-you-use/include-what-you-use/issues/679 indicates that clang may look for these headers at the path of `argv[0]`,didyou encounter something like this before @andreafioraldi @vanhauser-thc?

**Comment by tokatoka** [comment]
ty domenuk
is it only me that still cannot compile mozjpeg 😢

**Comment by domenukk** [comment]
The error looks like there is simply no clang? Seems odd though since it should be installed in github actions

**Comment by tokatoka** [comment]
yeah, and I can't compile libmozjpeg/libpng on my desktop either.
trying to pass `.add_arg("-fsanitize=fuzzer-no-link"), but clang is still calling `ld` 😓

**Comment by tokatoka** [comment]
I'm not 100% confident if I'm doing it right, but..

removing these, 
https://github.com/AFLplusplus/LibAFL/blob/4cafa8c2530ea10d402901f2bb0aa5d42bb50cde/libafl_cc/src/clang.rs#L168
https://github.com/AFLplusplus/LibAFL/blob/4cafa8c2530ea10d402901f2bb0aa5d42bb50cde/libafl_cc/src/clang.rs#L179

these `-Wl,-no-whole-archive`/`-Wl,--whole-archive` makes libafl_cc work.

**Comment by domenukk** [comment]
Does it just compile or actually work?

**Comment by tokatoka** [comment]
I can see the fuzzer getting new corpus entry, so I guess.. yes

**Comment by domenukk** [comment]
I think the examples that already have a `Makefile` should just call `make`

**Comment by domenukk** [comment]
@andreafioraldi maybe you can jump in here, I think your build skillz are better than mine...

**Comment by tokatoka** [comment]
from my observations, for libpng, the error is caused by building `png-fix-itxt` when linking with liblibfuzzer_libpng.a.
but if we use clang/gcc, the compiler does not have to link `png-fix-itxt` with `liblibfuzzer_libpng.a`.

so my guess is if we tune the libafl_cc so that the compiler call `link_staticlib` only when it's linking against the harness(, which has the LLVMTestOneInput symbol) would solve the issue

**Comment by andreafioraldi** [comment]
> from my observations, for libpng, the error is caused by building `png-fix-itxt` when linking with liblibfuzzer_libpng.a.
> but if we use clang/gcc, the compiler does not have to link `png-fix-itxt` with `liblibfuzzer_libpng.a`. (so it goes well)
> 
> so my guess is if we tune the libafl_cc so that the compiler call `link_staticlib` only when it's linking against the harness(, which has the LLVMTestOneInput symbol), that would solve the issue

yes true, i explained it to you on discord and then i fixed it. I don't understand the problem, for me it works

**Comment by tokatoka** [comment]
really... I tested on 3 different computers, none of them are successful to build libpng
am I missing the right commands to build libpng?
```
cd libpng-1.6.37
./configure
make CC=$(realpath ../target/release/libafl_cc) CXX=$(realpath ../target/release/libafl_cxx) -j `nproc`
```
these are still the right command to use to build libpng right?

can you reproduce the failure to build libpng @domenukk ?

**Comment by domenukk** [comment]
Seems to be dependency problems or something.
FWIW we could use the dockerfile like this:
https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action#creating-an-action-metadata-file

But it should be possible without, too

**Comment by domenukk** [comment]
For MacOS, we have this issue:
https://stackoverflow.com/questions/36523911/osx-homebrew-cmake-libpng-version-mismatch-issue

This is unrelated.

For Ubuntu, we have
```
 -- Performing Test INLINE_WORKS
-- Performing Test INLINE_WORKS - Failed
CMake Error at CMakeLists.txt:454 (message):
-- Configuring incomplete, errors occurred!
See also "/home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_libmozjpeg/mozjpeg-4.0.3/CMakeFiles/CMakeOutput.log".
See also "/home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_libmozjpeg/mozjpeg-4.0.3/CMakeFiles/CMakeError.log".
  Could not determine how to inline functions.
```
So probably our compiler wrapper does something wrong on this specific system?

**Comment by tokatoka** [comment]
should we skip building libpng/libmozjpeg on Apple for now?

**Comment by domenukk** [comment]
It builds locally for my apple so in theory it's possible (see answers in stackoverflow.com above) but skipping it may be fine too

File: scripts/build_all_fuzzers.sh [hunk 1]
```diff
@@ -5,6 +5,9 @@ cd "$SCRIPT_DIR/.."
 
 # TODO: This should be rewritten in rust, a Makefile, or some platform-independent language
 
+# Dependencies
+sudo apt install -y libc6-dev
```
Comment (domenukk) [inline scripts/build_all_fuzzers.sh:9]: This doesn't belong here imho

Comment (tokatoka) [reply scripts/build_all_fuzzers.sh]: yes should be in build_and_test.yml, and it's not needed

**Comment by tokatoka** [comment]
So it works for macos.... that is unexpected for me.
can you try building on ubuntu locally if you have a ubuntu computer domenuk?

File: fuzzers/libfuzzer_libmozjpeg/Makefile [hunk 2]
```diff
@@ -18,8 +18,8 @@ libafl_cxx: target/release/libafl_cxx
 libafl_cc: target/release/libafl_cxx
 
 mozjpeg-4.0.3-compiled: mozjpeg-4.0.3 libafl_cc
-	cd $(PWD)/mozjpeg-4.0.3 && cmake --disable-shared . -DCMAKE_C_COMPILER=$(abspath $(PWD)/../target/release/libafl_cc) -DCMAKE_CXX_COMPILER=$(abspath $(PWD)/../target/release/libafl_cxx) -G "Unix Makefiles"
```
Comment (andreafioraldi) [inline fuzzers/libfuzzer_libmozjpeg/Makefile:21]: @domenukk this is not how to write makefiles, it is broken

**Comment by andreafioraldi** [comment]
So, I can build on main but not on this branch, ergo the issue is in the code added here

File: fuzzers/libfuzzer_libmozjpeg/Makefile (hunk 2 above)
Comment (andreafioraldi) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: $(PWD) is correct

Comment (andreafioraldi) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: or better, PWD = $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))

File: fuzzers/libfuzzer_libpng_launcher/run.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
```
Comment (andreafioraldi) [inline fuzzers/libfuzzer_libpng_launcher/run.sh:1]: Why a shell script if we have the makefile?

**Comment by andreafioraldi** [comment]
No wtf main is broken, @domenukk did you touched stuffs in the compiler wrapper?

File: fuzzers/libfuzzer_libpng_launcher/run.sh (hunk 3 above)
Comment (tokatoka) [reply fuzzers/libfuzzer_libpng_launcher/run.sh:1]: yes, I should make Makefile for those fuzzers that are missing Makefile and remove run.sh

**Comment by domenukk** [comment]
Not as far as I know

File: fuzzers/libfuzzer_libmozjpeg/Makefile (hunk 2 above)
Comment (domenukk) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: `PWD` didn't work on Ubuntu in some occasions, according to ci. maybe that was in the sh file though, not sure

Comment (domenukk) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: Realpath doesn't exist on MacOS, don't use it

Comment (andreafioraldi) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: realpath is a makefile command, if make exists realpath exists. Btw macos has realpath it is part of coreutils

Comment (domenukk) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: It doesn't matter what it's part of, MacOS doesn't have it.
https://stackoverflow.com/questions/3572030/bash-script-absolute-path-with-os-x

Comment (domenukk) [reply fuzzers/libfuzzer_libmozjpeg/Makefile:21]: If it's a makefile command then we can use it in makefiles, but use $(pwd) in shell scripts, I can check.

**Comment by andreafioraldi** [comment]
@tokatoka can u test?

**Comment by tokatoka** [comment]
works perfect for libpng & mozjpeg, thanks!

I see the reason for `undefined reference to `LLVMFuzzerTestOneInput` now.
with EXT_FUNC, the symbol is undefined `U LLVMFuzzerTestOneInput`
with EXT_FUNC_IMPL, the symbol is weak defined, `W LLVMFuzzerTestOneInput`, making it successful to compile.

**Comment by tokatoka** [comment]
for macos the dependency issue still remains but we can just skip testing on Apple for now

**Comment by domenukk** [comment]
The ci now stops at
```bash
/usr/bin/ld: /home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_reachability/target/release/liblibfuzzer_libpng.a(libfuzzer_libpng-7be93c44005a6c57.libfuzzer_libpng.6dn10cv9-cgu.0.rcgu.o): in function `libfuzzer_libpng::fuzz':
/home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_reachability/src/lib.rs:72: undefined reference to `__libafl_target_list'
```

**Comment by tokatoka** [comment]
yes I'm looking at this now. If we are in a hurry we can skip this one and I'll make another pr later

**Comment by andreafioraldi** [comment]
> yes I'm looking at this now. If we are in a hurry we can skip this one and I'll make another pr later

good for me

**Comment by tokatoka** [comment]
```
/usr/bin/ld: /home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_reachability/target/release/liblibfuzzer_libpng.a(libfuzzer_libpng-7be93c44005a6c57.libfuzzer_libpng.6dn10cv9-cgu.0.rcgu.o): in function `libfuzzer_libpng::fuzz':
/home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_reachability/src/lib.rs:72: undefined reference to `__libafl_target_list'
```

I injected a pointer named __libafl_target_list into png.c
but it has also become `U __libafl_target_list` in liblibfuzzer_libpng.a

**Comment by tokatoka** [comment]
can I delete all run.sh and instead run `make short_test`?

**Comment by andreafioraldi** [comment]
> can I delete all run.sh and instead run `make short_test`?

y

**Comment by tokatoka** [comment]
```
/usr/bin/ld: /home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_reachability/target/release/liblibfuzzer_libpng.a(libfuzzer_libpng-7be93c44005a6c57.libfuzzer_libpng.6dn10cv9-cgu.0.rcgu.o): in function `libfuzzer_libpng::fuzz':
/home/runner/work/LibAFL/LibAFL/fuzzers/libfuzzer_reachability/src/lib.rs:72: undefined reference to `__libafl_target_list'
```
this problem can be solved by linking a object file like this one ⬇️  along with liblibfuzzer_libpng.a
```
#include<stddef.h>
__attribute__((weak, visibility("default"))) size_t *__libafl_target_list;
```
to make that symbol weakly defined (unfortunately __attribute__(weak) is available for nightly only.)
but, well, ... should be another PR.

**Comment by tokatoka** [comment]
if we are not on linux, we call
```
    else
        echo "[*] Building $fuzzer"
        cargo build || exit 1
        echo "[+] Done building $fuzzer"
```
instead to test building at least

**Comment by tokatoka** [comment]
it is good to go

**Comment by domenukk** [comment]
Now there is still
```
thread '<unnamed>' panicked at 'Failed to setup the restarter: ShuttingDown', src/lib.rs:64:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
   Compiling cfg-if v1.0.0
timeout: the monitored command dumped core
```
Which seems to be related to #191 but in any case the testcase is green, so lgtm :)

**Comment by tokatoka** [comment]
missed this one out
```
 timeout 11s ./"fuzzer_libpng" &
sleep 0.2
error: The following required arguments were not provided:
    --cores <CORES>

USAGE:
    fuzzer_libpng --cores <CORES>

For more information try --help
timeout 10s taskset -c 0 ./"fuzzer_libpng" >/dev/null 2>/dev/null &
timeout 10s taskset -c 1 ./"fuzzer_libpng" >/dev/null 2>/dev/null &
timeout 10s taskset -c 2 ./"fuzzer_libpng" >/dev/null 2>/dev/null &
```

libpng_launcher's usage is different, so I have to fix it

**Pull request merged**